src/main/java/software/amazon/nio/spi/s3/S3SeekableByteChannel.java

```diff
@@ -7,61 +7,58 @@
+import java.nio.file.OpenOption;
 import java.nio.file.StandardOpenOption;
+import java.util.Collections;
+import java.util.Set;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 
 public class S3SeekableByteChannel implements SeekableByteChannel {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(S3SeekableByteChannel.class);
+
     private long position;
     private final S3AsyncClient s3Client;
     private final S3Path path;
     private final ReadableByteChannel readDelegate;
-    //private final WritableByteChannel writeDelegate; // placeholder for when we implement writes.
+    private final S3WritableByteChannel writeDelegate;
 
+    private boolean closed;
     private long size = -1L;
 
-    private final Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+    protected S3SeekableByteChannel(S3Path s3Path, S3AsyncClient s3Client) throws IOException {
+        this(s3Path, s3Client, Collections.emptySet());
+    }
 
-    protected S3SeekableByteChannel(S3Path s3Path, S3AsyncClient s3Client, long startAt) throws IOException{
-        position = startAt;
+    protected S3SeekableByteChannel(S3Path s3Path, S3AsyncClient s3Client, Set<? extends OpenOption> options) throws IOException {
+        position = 0L;
         path = s3Path;
+        closed = false;
         this.s3Client = s3Client;
         s3Path.getFileSystem().registerOpenChannel(this);
 
         // later we will add a constructor that allows providing delegates for composition
-        logger.info("using S3ReadAheadByteChannel as read delegate for path '{}'", s3Path.toUri());
-        S3NioSpiConfiguration config = new S3NioSpiConfiguration();
-        readDelegate = new S3ReadAheadByteChannel(s3Path, config.getMaxFragmentSize(), config.getMaxFragmentNumber(), s3Client, this);
-    }
-
-    /**
-     * Open a new byte channel
-     * @param s3Path the path that the channel will read from
-     * @param startAt the byte offset to start at (0 is the beginning).
-     */
-    protected S3SeekableByteChannel(S3Path s3Path, long startAt) throws IOException {
-        this(s3Path, S3ClientStore.getInstance().getAsyncClientForBucketName(s3Path.bucketName()), startAt);
-    }
 
-    /**
-     * Equivalent to: {@code new S3SeekableByteChannel(s3Path, 0L)}
-     * @param s3Path the path to open the byte channel for.
-     */
-    protected S3SeekableByteChannel(S3Path s3Path) throws IOException {
-        this(s3Path, S3ClientStore.getInstance().getAsyncClientForBucketName(s3Path.bucketName()), 0L);
+        S3NioSpiConfiguration config = new S3NioSpiConfiguration();
+        if (options.contains(StandardOpenOption.WRITE)) {
+            LOGGER.info("using S3WritableByteChannel as write delegate for path '{}'", s3Path.toUri());
+            readDelegate = null;
+            writeDelegate = new S3WritableByteChannel(s3Path, s3Client, options);
+        } else {
```

Comment:
This assumes `StandardOpenOption.READ`? I think we should check that, error if READ and WRITE are selected. Probably error if other options are selected?